calc + upgrade calc

voki-postovoi-al opened the pull request.

voki-postovoi-al committed:
calc + upgrade calc

voki-postovoi-al committed:
change functions:
 add repeat call function if user enter not valid data.
cosmetic changes

voki-postovoi-al committed:
cosmetic changes

voki-baradulkin-a commented on python exercices/calc.py:
```diff
@@ -0,0 +1,67 @@
+        pass
+
+    def plus(self, a, b):
+        return print(a + b)
```
Мне кажется print() ничего ни возвращает...
по классие мы отдельно логируем, а потом возвращаем значение

voki-postovoi-al replied:
@voki-baradulkin-a тут соглашусь, print ничего не возвращает, по этому возврат будет None. 
Почитал по стандартизации ввёл доп. переменную result в каждом методе для  улучшения читабельности и простоты дебага

voki-baradulkin-a commented on python exercices/calc.py:
```diff
@@ -0,0 +1,67 @@
+
+
+    def plus(self, *args):
+        if len(args) == 1 and hasattr(args[0], '__iter__'):
```
Мне кажется решение через аргумент со значением по умолчанию 
def plus(a, b=None) 
немного изящнее. Ты по умолчанию 2 аргумент - это 2 числа, 1 аргумент - это список... если что-то пошло не так, кидаем ошибку... пусть сами разбираются)

voki-postovoi-al replied:
@voki-baradulkin-a поправил
Но небольшую проверку оставил

voki-baradulkin-a commented on python exercices/calc.py:
```diff
@@ -0,0 +1,67 @@
+            print('Invalid input')
+
+    def calculate_from_input(self, data=None):
+        pattern = r'^(-?\d+(\.\d+)?)([+\-*/])(-?\d+(\.\d+)?)$'
```
Мне кажется в (-?\d+(\.\d+)?) внутренние скобки лишние... тогда и парсинг будет более понятным
                a = float(parse_data.group(1))
                op = parse_data.group(2)
                b = float(parse_data.group(3))
Но это скорее про читаемость

voki-postovoi-al replied:
@voki-baradulkin-a переписал паттерн на r'^(-?\d+(?:\.\d+)?)([+\-*/])(-?\d+(?:\.\d+)?)$' что бы сократить количество групп с 5 до 3-х
Вторую часть с разделением оставил что бы проще читать было

voki-baradulkin-a commented on python exercices/calc.py:
```diff
@@ -0,0 +1,67 @@
+                    break
+            else:
+                print('not valid data, try again')
+                data = None
```
Тут data можно не трогать

voki-postovoi-al replied:
@voki-baradulkin-a Если оставить data без изменений то не верное значение которое вводил пользователь, будет без изменений, переменную надо переназначить что бы у пользователя заново запросило ввод.

voki-baradulkin-a commented on python exercices/calc.py:
```diff
@@ -0,0 +1,67 @@
+
+
+
+c = ModifiedCalc()
```
Вот эту штуку оборачивают в 
if __name__ == '__main__':
и в конце файла должна быть пустая строка

voki-postovoi-al replied:
@voki-baradulkin-a поправил

voki-postovoi-al committed:
Refactor calculator structure and clean up regex groups

Improve calculator input handling and fix minor logic bugs

Polish ModifiedCalc class and input parsing
